### PR TITLE
Accept new build flags when upgrading

### LINKED ISSUE #1699: Brew upgrade ignores all options
Currently `brew install -h` does not have a syntax for formula build options (--with/without-*):

```
brew install [--debug] [--env=std|super] [--ignore-dependencies] [--only-dependencies] [--cc=compiler] [--build-from-source] [--devel|--HEAD] [--keep-tmp] formula
brew install --interactive [--git] formula
```

This can cause confusion, because other commands use similar syntax but do not accept formula build options.

For example, `brew reinstall formula` (syntax from `brew reinstall -h`) accepts build options, but `brew upgrade [install-options] [--cleanup] [--fetch-HEAD] [formulae]` does not (despite specifying that it accepts `install-options`). `brew upgrade` exacerbates this irregularity by ignoring invalid arguments -- build options are simply ignored without warning. It can also be confusing that `brew upgrade` accepts `-s` but not build options.

I first thought that `brew upgrade` ignoring build options was a [bug](https://github.com/Homebrew/brew/issues/1656#issuecomment-267217170).

### PR DESCRIPTION
**Pull request by amyspark**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?
-----

This pull request fixes #1699 by accepting new build flags when upgrading a Formula

Please review, and thanks for your comments!

/cc @MikeMcQuaid

### CODE REVIEW

File: Library/Homebrew/cmd/upgrade.rb [hunk 1]
```diff
@@ -119,15 +119,19 @@ def upgrade_formula(f)
       tab = Tab.for_keg(keg)
     end
 
+    new_build_options = BuildOptions.new(Options.create(ARGV.flags_only), f.options)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/upgrade.rb]: Just go for `build_options` to be consistent with `reinstall` (or rename it there too if you prefer this name)

File: Library/Homebrew/cmd/upgrade.rb [hunk 2]
```diff
@@ -119,15 +119,19 @@ def upgrade_formula(f)
...
+
     fi = FormulaInstaller.new(f)
-    fi.options  = f.build.used_options
+    fi.options = new_build_options.used_options
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/upgrade.rb]: Match the `=` indentation from before.

File: Library/Homebrew/cmd/upgrade.rb [hunk 3]
```diff
@@ -119,15 +119,19 @@ def upgrade_formula(f)
...
+    fi.options |= f.build.used_options
     fi.options &= f.options
     fi.build_bottle = ARGV.build_bottle? || (!f.bottled? && f.build.build_bottle?)
     fi.installed_on_request = !ARGV.named.empty?
     fi.link_keg             = keg_was_linked if keg_had_linked_opt
     if tab
       fi.installed_as_dependency = tab.installed_as_dependency
       fi.installed_on_request  ||= tab.installed_on_request
+      fi.ignore_deps = tab.ignored_dependencies
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/upgrade.rb]: Let's pass on all `ignore_deps` additions; the current behaviour is desired.

File: Library/Homebrew/cmd/install.rb [hunk 4]
```diff
@@ -328,7 +328,6 @@ def install_formula(f)
     fi = FormulaInstaller.new(f)
     fi.options              = build_options.used_options
     fi.invalid_option_names = build_options.invalid_option_names
-    fi.ignore_deps          = ARGV.ignore_deps?
```
Comment (amyspark) [inline Library/Homebrew/cmd/install.rb]: 🤦‍♂️

**Comment by MikeMcQuaid** [comment]
Thanks again for the great PR, @amyspark!

**Pull request merged**